File: frontend/src/core/services/PictureService.ts
import { pictureAxios } from "@/core/axios";
import type { ResponseDto } from "../models/dto/ResponseDto";
import type { IPictureService } from "./iServices/IPictureService";
import type { PictureDto } from "../models/dto/PictureDto";

export class PictureService implements IPictureService {
  private static instance: PictureService;

  private constructor() {}

  public static getInstance(): PictureService {
    if (!PictureService.instance) {
      PictureService.instance = new PictureService();
    }
    return PictureService.instance;
  }

  async get(): Promise<ResponseDto> {
    try {
      const response = await pictureAxios.get<ResponseDto>("/get-pictures");

      if (!response.data.isSuccess) {
        throw new Error(response.data.message || "Failed to get pictures.");
      }

      return response.data;
    } catch (error: any) {
      throw error;
    }
  }

  async add(pictureDto: PictureDto): Promise<ResponseDto> {
    try {
      const response = await pictureAxios.put<ResponseDto>(
        "/add-picture",
        pictureDto
      );

      if (!response.data.isSuccess) {
        throw new Error(response.data.message || "Failed to upload picture.");
      }

      return response.data;
    } catch (error: any) {
      throw error;
    }
  }
}

File: Snp.Services.PictureApi/Snp.Services.PictureAlbumApi/Services/PictureService.cs
using AutoMapper;
using Microsoft.EntityFrameworkCore;
using Snp.Services.PictureAlbumApi.Data;
using Snp.Services.PictureAlbumApi.Exceptions;
using Snp.Services.PictureAlbumApi.Models;
using Snp.Services.PictureAlbumApi.Models.Dto;
using Snp.Services.PictureAlbumApi.Services.IServices;

namespace Snp.Services.PictureAlbumApi.Services
{
    public class PictureService : IPictureService
    {
        private readonly AppDbContext _db;
        private readonly IMapper _mapper;

        public PictureService(AppDbContext db, IMapper mapper)
        {
            _db = db;
            _mapper = mapper;
        }


        public async Task<string> AddAsync(PictureDto pictureDto)
        {
            if (await _db.Pictures.AnyAsync(p => p.File == pictureDto.File))
            {
                throw new PictureExisitsException();
            }

            var picture = _mapper.Map<Picture>(pictureDto);

            _db.Pictures.Add(picture);

            await _db.SaveChangesAsync();

            return picture.Id.ToString();
        }

        public async Task<IEnumerable<PictureListItemDto>> GetAsync()
        {
            var pictures = await GetAllPictures();
           
            return pictures;
        }

        private async Task<IEnumerable<PictureListItemDto>> GetAllPictures()
        {
            try
            {
                var pictures = await _db.Pictures.AsNoTracking().ToListAsync();

                return _mapper.Map<IEnumerable<PictureListItemDto>>(pictures);
            }
            catch (Exception ex)
            {
                throw new GetPicturesException(ex.Message);
            }
        }
    }
}
